pydantic/_internal/_generate_schema.py

```diff
@@ -247,11 +247,12 @@ def generate_field_schema(
         schema = self.generate_schema(field_info.annotation)
         schema = apply_annotations(schema, field_info.metadata)
 
+        schema = apply_validators(schema, validator_functions.get_field_decorators(name))
```

Comment:
There is a reason. Consider this pydantic-core example:

```python
from typing import Any
from pydantic_core import core_schema, SchemaValidator


def val(v: Any, info: core_schema.ValidationInfo) -> Any:
    return v


schema1 = core_schema.typed_dict_schema(
    fields={
        "a": core_schema.typed_dict_field(
            schema=core_schema.function_after_schema(
                function=val,
                schema=core_schema.with_default_schema(
                    core_schema.nullable_schema(core_schema.int_schema()), default=None
                ),
            ),
            required=False
        )
    }
)

v = SchemaValidator(schema1)
assert v.validate_python({}) == {}


schema2 = core_schema.typed_dict_schema(
    fields={
        "a": core_schema.typed_dict_field(
            schema=core_schema.with_default_schema(
                schema=core_schema.function_after_schema(
                    function=val, schema=core_schema.nullable_schema(core_schema.int_schema())
                ),
                default=None,
            ),
            required=False
        )
    }
)

v = SchemaValidator(schema2)
assert v.validate_python({}) == {"a": None}
```

`schema1` is what gets generated in main and `schema2` in this branch. Also corresponds to this example:

```python
class Model(BaseModel):
    x: Optional[int] = None

    @validator("x")
    def val_x(cls, v: Any) -> Any:
        return v

print(Model())
```

On `main` it gives an attribute error.

I didn't dig too deep but I'm guessing it's because the `required=False` option to the field only works if the inner schema supports a default value, which the `function` schemas do not.

Is this okay to keep then or do you consider this is a bug in `pydantic-core`? I think this could be it's own PR but since it depends on `@validator` it might be complicated to split out.

Reply:
Indeed the logic in `pydantic-core` seems hardcoded to check if the innner validator to `typed-dict-field` has a default value. Since this is not recursive and `function_after_schema` does not have a default value it just doesn't use one in the first case. It makes sense that switching the order fixes the problem.

Reply:
I think this is the same as #5045 - I was assuming we'd have to fix that in pydantic-core.

Please can you check if you're fix here resolves #5045, if so mark at issue to be fixed by this PR.

If not, I'll try to work on that soon - or alternatively, explain how I would fix it.

Reply:
Yep this does fix #5045!